File: tree.cpp
#include "tree.h"
#include "queue.h"
#include <iostream>
#include <algorithm>
using namespace std;

int getHeight(TNode* node) {
    return node ? node->height : 0;
}

void updateHeight(TNode* node) {
    if (node) {
        node->height = 1 + max(getHeight(node->left), getHeight(node->right));
    }
}

int getBalance(TNode* node) {
    return node ? getHeight(node->left) - getHeight(node->right) : 0;
}

TNode* createNode(string data) {
    TNode* node = new TNode;
    node->data = data;
    node->left = nullptr;
    node->right = nullptr;
    node->height = 1;
    return node;
}

// Правый поворот
TNode* rotateRight(TNode* y) {
    TNode* x = y->left;
    TNode* T2 = x->right;

    x->right = y;
    y->left = T2;

    updateHeight(y);
    updateHeight(x);

    return x;
}

// Левый поворот
TNode* rotateLeft(TNode* x) {
    TNode* y = x->right;
    TNode* T2 = y->left;

    y->left = x;
    x->right = T2;

    updateHeight(x);
    updateHeight(y);

    return y;
}

// Балансировка узла
TNode* balance(TNode* node) {
    if (!node) return nullptr;

    updateHeight(node);
    int bal = getBalance(node);

    // Left-Left
    if (bal > 1 && getBalance(node->left) >= 0) {
        return rotateRight(node);
    }

    // Left-Right
    if (bal > 1 && getBalance(node->left) < 0) {
        node->left = rotateLeft(node->left);
        return rotateRight(node);
    }

    // Right-Right
    if (bal < -1 && getBalance(node->right) <= 0) {
        return rotateLeft(node);
    }

    // Right-Left
    if (bal < -1 && getBalance(node->right) > 0) {
        node->right = rotateRight(node->right);
        return rotateLeft(node);
    }

    return node;
}

// Вставка узла
TNode* insertNode(TNode* node, string data, bool& inserted) {
    if (!node) {
        inserted = true;
        return createNode(data);
    }

    if (data < node->data) {
        node->left = insertNode(node->left, data, inserted);
    } else if (data > node->data) {
        node->right = insertNode(node->right, data, inserted);
    } else {
        // Дубликат - не вставляем
        return node;
    }

    return balance(node);
}

// Найти минимальный узел
TNode* findMin(TNode* node) {
    while (node && node->left) {
        node = node->left;
    }
    return node;
}

// Удаление узла
TNode* deleteNode(TNode* node, string data, bool& deleted) {
    if (!node) return nullptr;

    if (data < node->data) {
        node->left = deleteNode(node->left, data, deleted);
    } else if (data > node->data) {
        node->right = deleteNode(node->right, data, deleted);
    } else {
        deleted = true;

        // Узел с одним ребенком или без детей
        if (!node->left || !node->right) {
            TNode* temp = node->left ? node->left : node->right;
            delete node;
            return temp;
        }

        // Узел с двумя детьми
        TNode* temp = findMin(node->right);
        node->data = temp->data;
        node->right = deleteNode(node->right, temp->data, deleted);
        deleted = true;
    }

    return balance(node);
}

// Публичные функции
void CreateTree(BinaryTree* tree) {
    tree->root = nullptr;
    tree->size = 0;
}

void TPush(BinaryTree* tree, string data) {
    bool inserted = false;
    tree->root = insertNode(tree->root, data, inserted);
    if (inserted) {
        tree->size++;
    }
}

void TDelete(BinaryTree* tree, string data) {
    bool deleted = false;
    tree->root = deleteNode(tree->root, data, deleted);
    if (deleted) {
        tree->size--;
    }
}

bool TSearch(BinaryTree* tree, string data) {
    TNode* current = tree->root;
    while (current) {
        if (data == current->data) {
            return true;
        } else if (data < current->data) {
            current = current->left;
        } else {
            current = current->right;
        }
    }
    return false;
}

// Обход в ширину (level-order)
void PrintTree(TNode* root) {
    if (!root) {
        cout << "Дерево пустое" << endl;
        return;
    }

    Queue q;
    CreateQueue(&q);
    QPush(&q, to_string((long long)root));

    int level = 0;
    int nodesInLevel = 1;
    int nodesInNext = 0;

    cout << "Уровень " << level << ": ";

    while (q.size > 0) {
        string ptrStr = QPop(&q);
        TNode* current = (TNode*)(stoll(ptrStr));

        cout << current->data;
        nodesInLevel--;

        if (current->left) {
            QPush(&q, to_string((long long)current->left));
            nodesInNext++;
        }
        if (current->right) {
            QPush(&q, to_string((long long)current->right));
            nodesInNext++;
        }

        if (nodesInLevel > 0) {
            cout << " ";
        } else {
            cout << endl;
            if (nodesInNext > 0) {
                level++;
                cout << "Уровень " << level << ": ";
                nodesInLevel = nodesInNext;
                nodesInNext = 0;
            }
        }
    }

    DestroyQueue(&q);
}

// Прямой обход (Pre-order: корень -> левый -> правый)
void preOrder(TNode* node) {
    if (!node) return;
    cout << node->data << " ";
    preOrder(node->left);
    preOrder(node->right);
}

// Симметричный обход (In-order: левый -> корень -> правый)
void inOrder(TNode* node) {
    if (!node) return;
    inOrder(node->left);
    cout << node->data << " ";
    inOrder(node->right);
}

// Обратный обход (Post-order: левый -> правый -> корень)
void postOrder(TNode* node) {
    if (!node) return;
    postOrder(node->left);
    postOrder(node->right);
    cout << node->data << " ";
}

// Уничтожение дерева
void DestroyTree(TNode* node) {
    if (!node) return;
    DestroyTree(node->left);
    DestroyTree(node->right);
    delete node;
}

File: tree.go
package main

import "fmt"

type TNode struct {
	Data   string
	Left   *TNode
	Right  *TNode
	Height int
}

type BinaryTree struct {
	Root *TNode
	Size int
}

func getHeight(node *TNode) int {
	if node == nil {
		return 0
	}
	return node.Height
}

func updateHeight(node *TNode) {
	if node != nil {
		leftHeight := getHeight(node.Left)
		rightHeight := getHeight(node.Right)
		if leftHeight > rightHeight {
			node.Height = 1 + leftHeight
		} else {
			node.Height = 1 + rightHeight
		}
	}
}

func getBalance(node *TNode) int {
	if node == nil {
		return 0
	}
	return getHeight(node.Left) - getHeight(node.Right)
}

func createTNode(data string) *TNode {
	return &TNode{
		Data:   data,
		Left:   nil,
		Right:  nil,
		Height: 1,
	}
}

func rotateRight(y *TNode) *TNode {
	x := y.Left
	T2 := x.Right

	x.Right = y
	y.Left = T2

	updateHeight(y)
	updateHeight(x)

	return x
}

func rotateLeft(x *TNode) *TNode {
	y := x.Right
	T2 := y.Left

	y.Left = x
	x.Right = T2

	updateHeight(x)
	updateHeight(y)

	return y
}

func balance(node *TNode) *TNode {
	if node == nil {
		return nil
	}

	updateHeight(node)
	bal := getBalance(node)

	// Left-Left
	if bal > 1 && getBalance(node.Left) >= 0 {
		return rotateRight(node)
	}

	// Left-Right
	if bal > 1 && getBalance(node.Left) < 0 {
		node.Left = rotateLeft(node.Left)
		return rotateRight(node)
	}

	// Right-Right
	if bal < -1 && getBalance(node.Right) <= 0 {
		return rotateLeft(node)
	}

	// Right-Left
	if bal < -1 && getBalance(node.Right) > 0 {
		node.Right = rotateRight(node.Right)
		return rotateLeft(node)
	}

	return node
}

func insertNode(node *TNode, data string, inserted *bool) *TNode {
	if node == nil {
		*inserted = true
		return createTNode(data)
	}

	if data < node.Data {
		node.Left = insertNode(node.Left, data, inserted)
	} else if data > node.Data {
		node.Right = insertNode(node.Right, data, inserted)
	} else {
		return node
	}

	return balance(node)
}

func findMin(node *TNode) *TNode {
	for node != nil && node.Left != nil {
		node = node.Left
	}
	return node
}

func deleteNode(node *TNode, data string, deleted *bool) *TNode {
	if node == nil {
		return nil
	}

	if data < node.Data {
		node.Left = deleteNode(node.Left, data, deleted)
	} else if data > node.Data {
		node.Right = deleteNode(node.Right, data, deleted)
	} else {
		*deleted = true

		if node.Left == nil || node.Right == nil {
			if node.Left != nil {
				return node.Left
			}
			return node.Right
		}

		temp := findMin(node.Right)
		node.Data = temp.Data
		node.Right = deleteNode(node.Right, temp.Data, deleted)
		*deleted = true
	}

	return balance(node)
}

func CreateTree() *BinaryTree {
	return &BinaryTree{
		Root: nil,
		Size: 0,
	}
}

func (tree *BinaryTree) Push(data string) {
	inserted := false
	tree.Root = insertNode(tree.Root, data, &inserted)
	if inserted {
		tree.Size++
	}
}

func (tree *BinaryTree) Delete(data string) {
	deleted := false
	tree.Root = deleteNode(tree.Root, data, &deleted)
	if deleted {
		tree.Size--
	}
}

func (tree *BinaryTree) Search(data string) bool {
	current := tree.Root
	for current != nil {
		if data == current.Data {
			return true
		} else if data < current.Data {
			current = current.Left
		} else {
			current = current.Right
		}
	}
	return false
}

func (tree *BinaryTree) PrintTree() {
	if tree.Root == nil {
		fmt.Println("Дерево пустое")
		return
	}

	q := CreateQueue()
	q.Push(fmt.Sprintf("%p", tree.Root))

	level := 0
	nodesInLevel := 1
	nodesInNext := 0

	fmt.Printf("Уровень %d: ", level)

	for q.Size > 0 {
		ptrStr := q.Pop()
		var current *TNode
		fmt.Sscanf(ptrStr, "%p", &current)

		fmt.Print(current.Data)
		nodesInLevel--

		if current.Left != nil {
			q.Push(fmt.Sprintf("%p", current.Left))
			nodesInNext++
		}
		if current.Right != nil {
			q.Push(fmt.Sprintf("%p", current.Right))
			nodesInNext++
		}

		if nodesInLevel > 0 {
			fmt.Print(" ")
		} else {
			fmt.Println()
			if nodesInNext > 0 {
				level++
				fmt.Printf("Уровень %d: ", level)
				nodesInLevel = nodesInNext
				nodesInNext = 0
			}
		}
	}
}

func preOrder(node *TNode) {
	if node == nil {
		return
	}
	fmt.Print(node.Data, " ")
	preOrder(node.Left)
	preOrder(node.Right)
}

func inOrder(node *TNode) {
	if node == nil {
		return
	}
	inOrder(node.Left)
	fmt.Print(node.Data, " ")
	inOrder(node.Right)
}

func postOrder(node *TNode) {
	if node == nil {
		return
	}
	postOrder(node.Left)
	postOrder(node.Right)
	fmt.Print(node.Data, " ")
}